### PR TITLE
[SPARK-9458][SPARK-9469][SQL] Code generate prefix computation in sorting & moves unsafe conversion out of TungstenSort.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by rxin** [comment]
This also subsumes https://github.com/apache/spark/pull/7781

File: sql/core/src/test/scala/org/apache/spark/sql/execution/TungstenSortSuite.scala [hunk 1]
```diff
@@ -58,7 +58,7 @@ class UnsafeExternalSortSuite extends SparkPlanTest with BeforeAndAfterAll {
 
   // Test sorting on different data types
   for (
-    dataType <- DataTypeTestUtils.atomicTypes ++ Set(NullType)
+    dataType <- DataTypeTestUtils.atomicTypes.filter(_ != StringType) ++ Set(NullType)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/TungstenSortSuite.scala]: string is still failing

File: sql/core/src/main/scala/org/apache/spark/sql/execution/sort.scala [hunk 2]
```diff
@@ -97,59 +95,53 @@ case class ExternalSort(
...
+  override def canProcessSafeRows: Boolean = false
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputOrdering: Seq[SortOrder] = sortOrder
 
   override def requiredChildDistribution: Seq[Distribution] =
     if (global) OrderedDistribution(sortOrder) :: Nil else UnspecifiedDistribution :: Nil
 
-  protected override def doExecute(): RDD[InternalRow] = attachTree(this, "sort") {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/sort.scala]: Did you mean to remove this `attachTree`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 3]
```diff
@@ -37,4 +40,43 @@ case class SortOrder(child: Expression, direction: SortDirection)
...
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val childCode = child.child.gen(ctx)
+    val input = childCode.primitive
+    val DoublePrefixCmp = classOf[DoublePrefixComparator].getName
+
+    val (nullValue: Long, prefixCode: String) = child.child.dataType match {
+      case BooleanType =>
+        (Long.MinValue, s"$input ? 1L : 0L")
+      case _: IntegralType =>
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: Should we also support DateType and TimestampType here?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: Yea good idea to add.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/sort.scala (hunk 2 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/sort.scala]: Yea - other operators don't have it.

**Comment by JoshRosen** [comment]
Defining separate comparators for ascending and descending orderings is a nice trick.

LGTM.

**Comment by rxin** [comment]
Going to merge this since the test cases timed out at the very end testing Python.

**Pull request closed without merging**